feat(capitalize,toLowerCase,toUpperCase,uncapitalize): add `capitalize`, `toLowerCase`, `toUpperCase` and `uncapitalize`

lensbart opened the pull request:
Fixes https://github.com/remeda/remeda/issues/509

- [x] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [x] New methods added to `src/index.ts`
- [x] New methods added to `mapping.md`

lensbart commented:
@eranhirsch any chance that this can get a review? This PR has conflicts now, due to other PR’s that have been opened and merged in the meantime.

eranhirsch commented:
Hey, I'm sorry for the delays. The other PRs being merged are mine, and I'm sorry for the extra work they are incurring. I'm getting them done as quickly as possible to minimize these conflicts. The gist is that I'm reinforcing typescript and eslint (and prettier) in the project so that PRs get better feedback quicker, requiring fewer cycles and allowing me to review more PRs faster. Unfortunately, this delays reviewing the current PRs.

By the end of this week, those changes should be done, and then a simple merge from master should bring your environment up to speed You can follow the lint fixes and suggestions if any issues are raised. If you haven't already, you can always run `npx tsc`, `npm run lint` and `npx prettier . --write` to make sure you are getting the right signals from the CI.

eranhirsch reviewed (changes requested):
Hey, I merged the index file for you, can you pull this PR locally, run:
`npx prettier . --write` to bring styling up-to-date, and run `npm run lint` and fix any issues (I think it's just adding return types).

Also, please add dataLast implementations for all functions so they could be used in mappers and pipes, e.g.
```ts
pipe(
  stringData,
  map(toLowerCase()),
  ...
);
```

lensbart commented:
I rebased, re-formatted and added return types.

> Also, please add dataLast implementations for all functions so they could be used in mappers and pipes, e.g.
> 
> ```ts
> pipe(
>   stringData,
>   map(toLowerCase()),
>   ...
> );
> ```

Does this make sense? These functions all take just 1 argument, so they could safely be used in mappers and pipes like so:

```ts
pipe(
  stringData,
  map(toLowerCase),
  ...
);
```

eranhirsch commented:
We're moving away from 'headless' dataLast invocations for functions that don't take any parameters. They don't *always* work (like `first`), and that creates a confusing and inconsistent API where people might either add parenthesis to a headless function or not add parenthesis to a non-headless one and not get any meaningful errors to fix the issue.

See: https://github.com/remeda/remeda/pull/531

Unfortunately, we can't switch to a consistent format without breaking changes, so we are waiting to release Remeda V2 for this.

eranhirsch commented on src/capitalize.ts:
```diff
@@ -0,0 +1,14 @@
+/**
+ * A function equivalent to TypeScript’s built-in `Capitalize` utility type.
```
We don't need to mention "a function" because everything in Remeda is a function. Assuming people don't know Typescript's Capitalize util, it would be easier for people to get the full description for this function right here...

Suggestion (via chatgpt):
> Converts the first character of a given string to uppercase while maintaining the rest of the string as is.

eranhirsch commented on src/capitalize.ts:
```diff
@@ -0,0 +1,14 @@
+ * @category String
+ * @strict
+ */
+export const capitalize = <T extends string>(string: T): Capitalize<T> =>
```
To prevent any accidental confusion, let's avoid calling the variable `string` as it's also the type name. 
```suggestion
export const capitalize = <T extends string>(str: T): Capitalize<T> =>
```

eranhirsch commented on src/toLowerCase.test.ts:
```diff
@@ -0,0 +1,30 @@
+test("keep special characters as-is", () => {
+  expect(
+    toLowerCase(`
+😎
+/|\__👍
+/'\ `),
+  ).toEqual(`
+😎
+/|\__👍
+/'\ `);
+});
```
😎

eranhirsch commented on src/toLowerCase.test.ts:
```diff
@@ -0,0 +1,30 @@
+/* eslint-disable no-useless-escape */
```
put this above the line that needs it. If you need it for a block prefer wrapping that area with:

```ts
/* eslint-disable no-useless-escape */
...
...
...
/* eslint-enable no-useless-escape */
```

eranhirsch reviewed (changes requested):
Let's add native dataLast implementations

Brian-Pob commented:
Apologies for my ignorance, but what is the purpose of implementing functions like `toLoweCase` and `toUpperCase` when they already exist in native JavaScript? The only difference I can see is how you call them:

```js
R.toLowerCase("MyString");

"MyString".toLowerCase();

// alternatively
const mystr = "MyString";
mystr.toLowerCase();
```

Thanks.

lensbart commented:
@Brian-Pob

If some `variable` is typed as a string literal, the result of `toLowerCase(variable)` will be typed as the corresponding string literal, whereas `variable.toLowerCase()` will be typed as `string`.

Brian-Pob commented:
> @Brian-Pob
> 
> If some `variable` is typed as a string literal, the result of `toLowerCase(variable)` will be typed as the corresponding string literal, whereas `variable.toLowerCase()` will be typed as `string`.

I see. I get it now after reading your explanation and looking at the implementation. Thanks.

eranhirsch commented:
Also you get the added value of being able to use the dataLast version in pipes and mappers so you can write:
```ts
R.map(data, R.toLowerCase());
```

instead of:
```ts
data.map((tempName) => tempName.toLowerCase());
```

which itself becomes more useful once you start using pipes:
```ts
const result = R.pipe(
  someArrayOfNonStringData,
  R.filter(...),
  R.map(...a mapper that returns a string...),
  R.take(5),
  R.join("_"),
  R.toLowerCase(),
);
```

In the example above you will only run the filter and mapper on the first 5 objects that fulfil the mapper, you will only join 5 items, and then run lowercase just once on the output object.

Brodzko commented:
I'm also not sure what's the utility of `uncapitalize`. Are there many use-cases where you only want the first letter to lower-case?

lensbart commented:
@Brodzko `uncapitalize` can be used to convert PascalCase to camelCase. IMO the utility of `uncapitalize` is like the utility of [`Uncapitalize`](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html#uncapitalizestringtype). Since this utility type is part of the TypeScript API, in my view it makes sense to include the corresponding function in a TypeScript-specific utility library.

cjquines commented:
hey @lensbart — now that we released v2, we should point this PR to `main` instead of `master`. do you still intend to work on this?

cenobitedk commented:
Hi @lensbart would love to get these features in remeda and ditch lodash. Do you have time to update the PR?

lensbart commented:
Apologies for the late response!

I might have time in a couple of weeks when work slows down, but (after having fixed a couple of rounds of feedback here) I just copied the code to my repo to be done with it, so it’s not a high priority for me unfortunately.

Feel free to push to this branch (or inline the code yourself) if you want something sooner. Thanks

eranhirsch commented:
This PR was closed because the `master` branch it was based on is no longer correct. If you still want to finish this PR, please rebase it on top of `main` and re-open it.
